server/src/storage/compressed_data_buffer.rs

```diff
@@ -0,0 +1,169 @@
+/// functionality for appending segments and saving all segments to a single Apache Parquet file.
+pub(super) struct CompressedDataBuffer {
+    /// Compressed segments that make up the compressed data in the [`CompressedDataBuffer`].
+    compressed_segments: Vec<RecordBatch>,
+    /// Continuously updated total sum of the size of the compressed segments.
+    pub(super) size_in_bytes: usize,
+}
+
+impl CompressedDataBuffer {
+    pub(super) fn new() -> Self {
+        Self {
+            compressed_segments: Vec::new(),
+            size_in_bytes: 0,
+        }
+    }
+
+    /// Append `compressed_segments` to the [`CompressedDataBuffer`] and return the size of
+    /// `compressed_segments` in bytes. It is assumed that `compressed_segments` is sorted by time.
+    pub(super) fn append_compressed_segments(&mut self, compressed_segments: RecordBatch) -> usize {
+        let segment_size = Self::get_size_of_compressed_segments(&compressed_segments);
+
+        self.compressed_segments.push(compressed_segments);
+        self.size_in_bytes += segment_size;
+
+        segment_size
+    }
+
+    /// If the compressed segments are successfully saved to an Apache Parquet file return [`Ok`],
+    /// otherwise return [`IOError`].
+    pub(super) fn save_to_apache_parquet(
+        &mut self,
+        folder_path: &Path,
+        compressed_schema: &CompressedSchema,
+    ) -> Result<(), IOError> {
+        debug_assert!(
+            !self.compressed_segments.is_empty(),
+            "Cannot save CompressedDataBuffer with no data."
+        );
+
+        // Combine the compressed segments into a single RecordBatch.
+        let batch =
+            compute::concat_batches(&compressed_schema.0, &self.compressed_segments).unwrap();
+
+        // Create the folder structure if it does not already exist.
+        fs::create_dir_all(folder_path)?;
+
+        // Create a new file that includes the start timestamp of the first segment in batch, the
+        // end timestamp of the last segment in batch, the minium value stored in batch, and the
+        // maximum value stored in batch as the file name to efficiently pruning files that only
+        // contains data points with timestamps and values that is not relevant for a given query.
+        let file_name = storage::create_time_and_value_range_file_name(&batch);
+        let file_path = folder_path.join(file_name);
+        StorageEngine::write_batch_to_apache_parquet_file(batch, file_path.as_path())
+            .map_err(|error| IOError::new(Other, error.to_string()))?;
+
+        Ok(())
+    }
+
+    /// Return the size in bytes of `compressed_segments`.
+    fn get_size_of_compressed_segments(compressed_segments: &RecordBatch) -> usize {
+        let mut total_size: usize = 0;
+
+        // Compute the total number of bytes of memory used by the columns.
+        for column in compressed_segments.columns() {
+            total_size += column.get_array_memory_size()
```

Comment:
Previously there was a TODO here that mentioned that we should look into how this is calculated internally. Has this been removed because it has been done or because you did not find it necessary. It is fine either way, just curious.

Reply:
Restored TODO.